platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+            Some(ErrorKind::DataStart)
+        } else if mint.m_dee_int().bit_is_set() {
+            Some(ErrorKind::DataEnd)
+        } else if idst.des_unavl_int().bit_is_set() || idst.fatal_berr_int().bit_is_set() {
+            Some(ErrorKind::Dma)
+        } else {
+            None
+        }
+    }
+
+    fn advance_isr(&mut self, smhc: &smhc::RegisterBlock, num: u8) {
+        tracing::trace!(state = ?self.state, smhc = num, "SMHC{num} interrupt");
+
+        let mut needs_wake = false;
+
+        self.err = Self::error_status(smhc);
+        if let Some(err) = self.err {
+            // Clear all interrupt bits
+            smhc.smhc_rintsts.write(|w| unsafe { w.bits(0xFFFF_FFFF) });
+            smhc.smhc_idst.write(|w| unsafe { w.bits(0x3FF) });
+
+            needs_wake = true;
+        }
+
+        self.state = match self.state {
+            State::Idle => State::Idle,
+            State::WaitForCommand => {
+                if smhc.smhc_mintsts.read().m_cc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.cc().set_bit());
+                    match self.op {
+                        SmhcOp::None => State::Idle,
+                        SmhcOp::Control => {
+                            needs_wake = true;
+                            State::Idle
+                        }
+                        SmhcOp::Read { .. } | SmhcOp::Write { .. } => State::WaitForDma,
+                    }
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDma => {
+                // TODO: better way to check for RX and TX,
+                // *normal interrupt summary* does not seem to work
+                if smhc.smhc_idst.read().bits() != 0 {
+                    smhc.smhc_idst.write(|w| {
+                        w.nor_int_sum().set_bit();
+                        w.rx_int().set_bit();
+                        w.tx_int().set_bit();
+                        w
+                    });
+                    State::WaitForDataTransfer
+                } else {
+                    self.state
+                }
+            }
+            State::WaitForDataTransfer => {
+                if smhc.smhc_mintsts.read().m_dtc_int().bit_is_set() {
+                    smhc.smhc_rintsts.write(|w| w.dtc().set_bit());
+                    match &mut self.op {
+                        SmhcOp::None | SmhcOp::Control => State::Idle,
+                        SmhcOp::Read {
+                            auto_stop,
+                            buf,
+                            cnt,
+                        }
+                        | SmhcOp::Write {
+                            auto_stop,
+                            buf,
+                            cnt,
+                        } => {
+                            tracing::trace!("setting buf len: {cnt}");
+                            // TODO: Safety?
+                            unsafe { buf.as_vec_mut().set_len(*cnt) };
```

Comment:
One downside of using a `FixedVec`: having to manually set the length after you read the data.